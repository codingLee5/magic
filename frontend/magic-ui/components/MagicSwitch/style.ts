import { createStyles } from "antd-style"

export const useStyles = createStyles(({ prefixCls, token, css }) => {
	return {
		magicSwitch: css`
			background-color: ${token.magicColorUsages.fill[0]};
			.${prefixCls}-switch-handle {
				&::before {
					box-shadow: 0px 0px 1px 0px rgba(0, 0, 0, 0.3),
						0px 4px 6px 0px rgba(0, 0, 0, 0.1);
					border: 1px solid ${token.magicColorUsages.border};
				}
			}
			&:hover {
				background-color: ${token.magicColorScales.grey[1]} !important;
			}
			&.${prefixCls}-switch-loading {
				.${prefixCls}-switch-handle {
					&::before {
						background-color: transparent;
						border: none;
						box-shadow: none;
					}
					.${prefixCls}-switch-loading-icon {
						color: #fff;
					}
				}
			}
			&[disabled] {
				.${prefixCls}-switch-inner {
					background-color: #d3dffb;
				}
			}
			&[aria-checked="true"] {
				.${prefixCls}-switch-inner {
					background: #315cec;
				}
				&:hover {
					.${prefixCls}-switch-inner {
						background-color: #2447c8;
					}
				}
			}
		`,
	}
})
