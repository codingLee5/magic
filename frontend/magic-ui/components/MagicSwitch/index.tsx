import { Switch, type SwitchProps } from "antd"
import { useStyles } from "./style"

export type MagicSwitchProps = SwitchProps

function MagicSwitch({ className, ...props }: MagicSwitchProps) {
	const { styles, cx } = useStyles()

	return <Switch className={cx(styles.magicSwitch, className)} {...props} />
}

export default MagicSwitch
