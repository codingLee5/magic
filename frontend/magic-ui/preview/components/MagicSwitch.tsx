import React from "react"
import { Space } from "antd"
import MagicSwitch from "../../components/MagicSwitch"
import ComponentDemo from "./Container"

const SwitchDemo: React.FC = () => {
	return (
		<div>
			<ComponentDemo title="基础开关" description="最基本的开关组件" code="<MagicSwitch />">
				<Space>
					<MagicSwitch />
					<MagicSwitch defaultChecked />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="开关尺寸"
				description="支持不同尺寸的开关"
				code="size: 'default' | 'small'"
			>
				<Space>
					<MagicSwitch size="default" />
					<MagicSwitch size="small" />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="开关状态"
				description="支持加载和禁用状态"
				code="loading | disabled"
			>
				<Space>
					<MagicSwitch loading />
					<MagicSwitch size="small" loading />
					<MagicSwitch disabled />
					<MagicSwitch defaultChecked disabled />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="带文字的开关"
				description="可以自定义开关两端的文字"
				code="checkedChildren | unCheckedChildren"
			>
				<Space direction="vertical">
					<MagicSwitch checkedChildren="开" unCheckedChildren="关" />
					<MagicSwitch checkedChildren="ON" unCheckedChildren="OFF" />
					<MagicSwitch checkedChildren="✅" unCheckedChildren="❌" />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="受控开关"
				description="通过checked和onChange控制开关状态"
				code="checked | onChange"
			>
				<Space>
					<MagicSwitch
						checked={true}
						onChange={(checked) => console.log("开关状态:", checked)}
					/>
					<MagicSwitch
						checked={false}
						onChange={(checked) => console.log("开关状态:", checked)}
					/>
				</Space>
			</ComponentDemo>
		</div>
	)
}

export default SwitchDemo
