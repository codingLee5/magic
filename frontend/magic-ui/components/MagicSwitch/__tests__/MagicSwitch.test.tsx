import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicSwitch from "../index"

const renderWithTheme = (component: React.ReactElement) =>
	render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)

describe("MagicSwitch", () => {
	it("应该正常渲染", () => {
		renderWithTheme(<MagicSwitch />)
		expect(screen.getByRole("switch")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("默认开关快照", () => {
			const { asFragment } = renderWithTheme(<MagicSwitch />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("开启状态开关快照", () => {
			const { asFragment } = renderWithTheme(<MagicSwitch defaultChecked />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("禁用状态开关快照", () => {
			const { asFragment } = renderWithTheme(<MagicSwitch disabled />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("禁用且开启状态开关快照", () => {
			const { asFragment } = renderWithTheme(<MagicSwitch disabled defaultChecked />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("小尺寸开关快照", () => {
			const { asFragment } = renderWithTheme(<MagicSwitch size="small" />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带标签开关快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicSwitch checkedChildren="开" unCheckedChildren="关" />,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
