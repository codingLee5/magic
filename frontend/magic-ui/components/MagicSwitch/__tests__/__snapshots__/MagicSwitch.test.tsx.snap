// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicSwitch > 快照测试 > 小尺寸开关快照 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="magic-switch magic-switch-small acss-17xfzng css-dev-only-do-not-override-yotw4n css-var-r5"
    role="switch"
    type="button"
  >
    <div
      class="magic-switch-handle"
    />
    <span
      class="magic-switch-inner"
    >
      <span
        class="magic-switch-inner-checked"
      />
      <span
        class="magic-switch-inner-unchecked"
      />
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicSwitch > 快照测试 > 带标签开关快照 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="magic-switch acss-17xfzng css-dev-only-do-not-override-yotw4n css-var-r6"
    role="switch"
    type="button"
  >
    <div
      class="magic-switch-handle"
    />
    <span
      class="magic-switch-inner"
    >
      <span
        class="magic-switch-inner-checked"
      >
        开
      </span>
      <span
        class="magic-switch-inner-unchecked"
      >
        关
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicSwitch > 快照测试 > 开启状态开关快照 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="magic-switch acss-17xfzng css-dev-only-do-not-override-yotw4n css-var-r2 magic-switch-checked"
    role="switch"
    type="button"
  >
    <div
      class="magic-switch-handle"
    />
    <span
      class="magic-switch-inner"
    >
      <span
        class="magic-switch-inner-checked"
      />
      <span
        class="magic-switch-inner-unchecked"
      />
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicSwitch > 快照测试 > 禁用且开启状态开关快照 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="magic-switch acss-17xfzng css-dev-only-do-not-override-yotw4n css-var-r4 magic-switch-checked magic-switch-disabled"
    disabled=""
    role="switch"
    type="button"
  >
    <div
      class="magic-switch-handle"
    />
    <span
      class="magic-switch-inner"
    >
      <span
        class="magic-switch-inner-checked"
      />
      <span
        class="magic-switch-inner-unchecked"
      />
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicSwitch > 快照测试 > 禁用状态开关快照 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="magic-switch acss-17xfzng css-dev-only-do-not-override-yotw4n css-var-r3 magic-switch-disabled"
    disabled=""
    role="switch"
    type="button"
  >
    <div
      class="magic-switch-handle"
    />
    <span
      class="magic-switch-inner"
    >
      <span
        class="magic-switch-inner-checked"
      />
      <span
        class="magic-switch-inner-unchecked"
      />
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicSwitch > 快照测试 > 默认开关快照 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="magic-switch acss-17xfzng css-dev-only-do-not-override-yotw4n css-var-r1"
    role="switch"
    type="button"
  >
    <div
      class="magic-switch-handle"
    />
    <span
      class="magic-switch-inner"
    >
      <span
        class="magic-switch-inner-checked"
      />
      <span
        class="magic-switch-inner-unchecked"
      />
    </span>
  </button>
</DocumentFragment>
`;
